Added a feature to enable glowroot in tomcat

cgbautista opened the pull request:
Added a feature to support an open source APM for the tomcat process: glowroot.
There are 3 new parameters to the `start` function:
 - `--glowroot`: it downloads the latest version of glowroot and deploys it under /opt/glowroot. Activates the APM in tomcat and allos incoming connections.
 -  `--glowroot-zip=_FILE_`: instead of allowing to dectect and download the latest version available, it can specify a zip file with any custom version of glowroot, in case the latest version wouldn't work with the tomcat version running or had any issues. It is ignored if `--glowroot` is specified.
 - `--glowroot-port=_PORT_`: it allows to modify the default glowroot port exposed from 4000 to the desired one. Will not do anything if neither `--glowroot` nor `--glowroot-zip` are specified.

When enabled, the exposed endpoint can be reached at `http://localhost:4000` by default.

Ramon-Jimenez committed:
Merge pull request #139 from EyeSeeTea/development

Release 1.15.0

Ramon-Jimenez committed:
Merge pull request #147 from EyeSeeTea/development

Release 0.15.1

cgbautista committed:
Added a feature to enable an open-source APM: glowroot to the tomcat service. It downloads the latest version (--glowroot) or you can specify an specific zip file that you already have downloaded (--glowroot-zip), by default it will listen on port 4000, but you can change it with --glowroot-port (only if glowroot is enabled by any of the other two ways). Latest version is prioritized over an specific file

tokland commented on src/d2_docker/config/dhis2-core-entrypoint.sh:
```diff
@@ -49,6 +52,18 @@ if [ "$(id -u)" = "0" ]; then
         rm -v $WARFILE # just to save space
     fi
 
+    if [ -f $GLOWROOT_ZIP ] && [ ! -d $GLOWROOT_DIR ] ; then
```
I'd move the block to some function to keep this main function short and declarative.

tokland commented on src/d2_docker/config/dhis2-core-entrypoint.sh:
```diff
@@ -49,6 +52,18 @@ if [ "$(id -u)" = "0" ]; then
     fi
 
+    if [ -f $GLOWROOT_ZIP ] && [ ! -d $GLOWROOT_DIR ] ; then
+        unzip -q $GLOWROOT_ZIP -d /opt/ || [ $? -le 1 ]
```
That's a bit cryptic. I gather a code with value 1 for unzip is non-terminal, and we can continue. I'd refactor it so the intent is more clear, something like:

```
status=0
unzip -q "$GLOWROOT_ZIP" -d /opt/ || status=$?
if [ $status -gt 1 ]; then
  exit $status
fi
```

tokland commented on src/d2_docker/utils.py:
```diff
@@ -270,11 +285,30 @@ def run_docker_compose(
+    glowroot_path=None
+
+    if args[0] == "up":
+        glowroot_file = tempfile.NamedTemporaryFile(delete=False, prefix="glowroot_", suffix=".zip", dir="/tmp")
```
I'd move the glowroot block to a separate function. If you feel like it, don't be afraid of creating a new glowroot.py function that includes all the code regarding the feature. Jorge is wary of big fat utils files like the one I created here, and for a good reason!

tokland commented on src/d2_docker/utils.py:
```diff
@@ -296,13 +330,18 @@ def run_docker_compose(
         # Add ROOT_PATH from environment (required when run inside a docker)
         ("ROOT_PATH", ROOT_PATH),
         ("PSQL_ENABLE_QUERY_LOGS", "") if not enable_postgres_queries_logging else None,
+        ("GLOWROOT_PORT", "{}:4000".format(glowroot_port) if glowroot_port else "4000:4000") if (glowroot or glowroot_zip) else ("GLOWROOT_PORT", None),
```
Let's move this 4000 to a variable at the top of the file to make it more visible.

tokland commented on README.md:
```diff
@@ -92,6 +94,9 @@ Some notes:
 - Use option `--debug-port=PORT` to specify the debug port of the Tomcat process.
+- Use option `--glowroot-port=PORT` to specify the APM glowroot port of the Tomcat process.
+- Use option `--glowroot` to use the latest version of glowroot in the Tomcat process.
+- Use option `--glowroot-zip=FILE` to specify the zip file with a version of glowroot to run in the Tomcat process.
```
A proposal: Instead of adding the three options in this section of the README, create a new one dedicated to Glowroot. Then we can explain what an APM and Glowroot is. After that, we may show the d2-docker start with the options. What do you think?

tokland reviewed (commented):
Some minor in-line comments:

cgbautista committed:
Code cleanup

tokland commented:
A minor thing: when not using Glowroot, we see in the logs: "core-1     | warning [/opt/glowroot.zip]: zipfile is empty". The user may think something is wrong. Can we refactor to hide this or show a more descriptive log entry?

cgbautista committed:
Avoid warning output from unzip

cgbautista committed:
status was always 0 because incorrect placement of parenthesis

cgbautista committed:
Fixed an error when `d2-docker logs` was used.

cgbautista committed:
Added an option in the docker-compose.yml to ensure there is no "search"
domain in the /etc/resolv.conf copied onto the containers
This avoids the `core` container trying to determine when the `data`
container has finished its execution via `curl` to get responses from
any `data.search_domain` machine that might exist.

cgbautista committed:
Moved 4 functions inside a condition to avoid reseting container folders
to the original content (it didn't erase contents, but overwrote any
change on those folders with the files' previous version)

cgbautista committed:
extracted setup_tomcat, since it will copy any dhis.conf or server.xml
that is passed as arguments

cgbautista committed:
Fix in tabs

cgbautista committed:
After stopping the running containers, the command "docker compose ps
-q" did not return any container, so a "-a" parameter is added to ensure
all processes (even stopped) are returned

tokland commented on src/d2_docker/config/dhis2-core-entrypoint.sh:
```diff
@@ -46,14 +46,15 @@ wait_for_data_container_to_finish_copy() {
     if [ -f $GLOWROOT_ZIP ] && [ ! -d $GLOWROOT_DIR ] ; then
-        unzip -q $GLOWROOT_ZIP -d /opt/ || [ $? -le 1 ]
+        status=0
+	output=$(unzip -q "$GLOWROOT_ZIP" -d /opt/ 2>&1) || status=$?
```
[formatting] I'd say to test with vscode+some extension (ideally with some .editconfig so we all get the same)

tokland commented on src/d2_docker/glowroot.py:
```diff
@@ -0,0 +1,46 @@
+
+def get_glowroot_zip(command, glowroot_zip, glowroot):
+    logger = utils.logger
+    glowroot_path=None
```
formatting

tokland commented on src/d2_docker/glowroot.py:
```diff
@@ -0,0 +1,46 @@
+def get_glowroot_zip(command, glowroot_zip, glowroot):
+    logger = utils.logger
+    glowroot_path=None
+    if isinstance(command, list) and command[0] == "up":
```
keeping the spirit of immutability, I'd add an explicit else to better manage glowroot_path.

tokland commented on README.md:
```diff
@@ -381,3 +378,17 @@ $ cp flaskenv.secret ~/.config/d2-docker/
+When starting a container, there are a few options to enable glowroot on the Tomcat process:
+- Use option `--glowroot-port=PORT` to specify the APM glowroot port of the Tomcat process.
+- Use option `--glowroot` to use the latest version of glowroot in the Tomcat process.
+- Use option `--glowroot-zip=FILE` to specify the zip file with a version of glowroot to run in the Tomcat process.
```
What's the default port if not specified?

tokland replied:
It's not clear how to the options form a group. I guess --glowroot | glowroot-zip, and --glowroot-port applies to both.

tokland commented:
@cgbautista can you link this PR to its task in clickup?

cgbautista committed:
Fixed formatting and Readme

tokland commented:
@cgbautista can you link this PR (in the description) to some task in clickup?

(this one? https://app.clickup.com/t/8699xm4qd)

ifoche committed:
Merge pull request #153 from EyeSeeTea/fix/rm_command_did_not_clear_stopped_containers

"d2-docker rm" not clearing up containers

ifoche committed:
Merge pull request #152 from EyeSeeTea/fix/containers_should_keep_new_apps

[FIX] -k option should keep apps, dataValues, documents, etc. upon restart

ifoche committed:
Merge pull request #151 from EyeSeeTea/fix/dns_search_domains

[Fix] Core container might get responses from other servers named "data"
